recipes/openai_textcat.py

```diff
@@ -0,0 +1,535 @@
+        self.openai_max_tokens = openai_max_tokens
+        self.openai_timeout_s = openai_timeout_s
+        self.openai_n = openai_n
+
+    def __call__(
+        self,
+        stream: Iterable[Dict],
+        *,
+        nlp: Language,
+        batch_size: int,
+        negative_bias: float = 1.0,
+    ) -> Iterable[Dict]:
+        if self.segment:
+            stream = prodigy.components.preprocess.split_sentences(nlp, stream)
+
+        stream = self.stream_suggestions(stream, batch_size=batch_size)
+        stream = self.format_suggestions(stream, nlp=nlp)
+        stream = self.filter_suggestions(stream, negative_bias=negative_bias)
+        return stream
+
+    def update(self, examples: Iterable[Dict]) -> float:
+        for eg in examples:
+            if PromptExample.is_flagged(eg):
+                self.add_example(PromptExample.from_prodigy(eg, self.label))
+        return 0.0
+
+    def add_example(self, example: PromptExample) -> None:
+        """Add an example for use in the prompts. Examples are pruned to the most recent max_examples."""
+        if self.max_examples:
+            self.examples.append(example)
+        if len(self.examples) >= self.max_examples:
+            self.examples = self.examples[-self.max_examples :]
+
+    def stream_suggestions(
+        self, stream: Iterable[Dict], batch_size: int
+    ) -> Iterable[Dict]:
+        """Get zero-shot or few-shot suggested TextCat annotations from OpenAI.
+
+        Given a stream of input examples, we define a prompt, get a response from OpenAI,
+        and yield each example with their predictions to the output stream.
+        """
+        for batch in _batch_sequence(stream, batch_size):
+            prompts = [
+                self._get_textcat_prompt(
+                    eg["text"], label=self.label, examples=self.examples
+                )
+                for eg in batch
+            ]
+            responses = self._get_textcat_response(prompts, delay=0)
+            for eg, prompt, response in zip(batch, prompts, responses):
+                if self.verbose:
+                    rich.print(Panel(prompt, title="Prompt to OpenAI"))
+                eg["openai"] = {"prompt": prompt, "response": response}
+                if self.verbose:
+                    rich.print(Panel(response, title="Response from OpenAI"))
+                yield eg
+
+    def format_suggestions(
+        self, stream: Iterable[Dict], *, nlp: Language
+    ) -> Iterable[Dict]:
+        """Parse the examples in the stream and set up labels
+        to display in the Prodigy UI.
+        """
+        stream = prodigy.components.preprocess.add_tokens(nlp, stream, skip=True)  # type: ignore
+        for example in stream:
+            example = copy.deepcopy(example)
+            response = self._parse_response(example["openai"]["response"])
+            example["answer"] = response["answer"] == "accept"
+            example["meta"]["reason"] = response["reason"]
```

Comment:
I tried running the base recipe and I think I hit a bug. I have a `doubt.jsonl` file 

```
python -m prodigy textcat.openai.suggest openai-text-demo doubt.jsonl excitement  -F recipes/openai_textcat.py
```

This resulted in this error. 

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/prodigy/__main__.py", line 61, in <module>
    controller = recipe(*args, use_plac=True)
  File "cython_src/prodigy/core.pyx", line 339, in prodigy.core.recipe.recipe_decorator.recipe_proxy
  File "cython_src/prodigy/core.pyx", line 366, in prodigy.core._components_to_ctrl
  File "cython_src/prodigy/core.pyx", line 125, in prodigy.core.Controller.__init__
  File "cython_src/prodigy/components/feeds.pyx", line 104, in prodigy.components.feeds.Feed.__init__
  File "cython_src/prodigy/components/feeds.pyx", line 150, in prodigy.components.feeds.Feed._init_stream
  File "cython_src/prodigy/components/stream.pyx", line 107, in prodigy.components.stream.Stream.__init__
  File "cython_src/prodigy/components/stream.pyx", line 58, in prodigy.components.stream.validate_stream
  File "recipes/openai_textcat.py", line 209, in filter_suggestions
    for example in stream:
  File "recipes/openai_textcat.py", line 202, in format_suggestions
    example["meta"]["reason"] = response["reason"]
KeyError: 'meta'
```

I was able to fix this by adding this to line 202:

```python
if "meta" not in example:
    example["meta"] = {}
```

Reply:
The same error pops up in the multilabel file.